Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +95,349 @@ extension CertificatePolicy {
+                                   diagnosticsEngine: diagnosticsEngine, callbackQueue: callbackQueue, callback: callback)
+        } else {
+            wrappedCallback(.success(()))
+        }
     }
+    #endif
 }
 
+#if os(Linux) || os(Windows)
+private let ocspClient = BoringSSLOCSPClient()
+
+private struct BoringSSLOCSPClient {
+    private let resultCache = ThreadSafeKeyValueStore<CacheKey, CacheValue>()
+
+    private let cacheTTL: DispatchTimeInterval
+
+    init(cacheTTL: DispatchTimeInterval = .seconds(300)) {
+        self.cacheTTL = cacheTTL
+    }
+
+    func checkStatus(certificate: Certificate,
+                     issuer: Certificate,
+                     anchorCerts: [Certificate]?,
+                     httpClient: HTTPClient,
+                     diagnosticsEngine: DiagnosticsEngine,
+                     callbackQueue: DispatchQueue,
+                     callback: @escaping (Result<Void, Error>) -> Void) {
+        let wrappedCallback: (Result<Void, Error>) -> Void = { result in callbackQueue.async { callback(result) } }
+
+        let ocspURLs = certificate.withUnsafeMutablePointer { CCryptoBoringSSL_X509_get1_ocsp($0) }
+        defer { CCryptoBoringSSL_sk_OPENSSL_STRING_free(ocspURLs) }
+
+        let ocspURLCount = CCryptoBoringSSL_sk_OPENSSL_STRING_num(ocspURLs)
+        // Nothing to do if no OCSP URLs. Use `supportsOCSP` to require OCSP support if needed.
+        guard ocspURLCount > 0 else { return wrappedCallback(.success(())) }
+
+        // Construct the OCSP request
+        let digest = CCryptoBoringSSL_EVP_sha1()
+        let certid = certificate.withUnsafeMutablePointer { certPtr in
+            issuer.withUnsafeMutablePointer { issPtr in
+                OCSP_cert_to_id(digest, certPtr, issPtr)
+            }
+        }
+        let request = OCSP_REQUEST_new()
+        defer { OCSP_REQUEST_free(request) }
+
+        guard OCSP_request_add0_id(request, certid) != nil else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        // Write the request binary to memory bio
+        let bio = CCryptoBoringSSL_BIO_new(CCryptoBoringSSL_BIO_s_mem())
+        defer { CCryptoBoringSSL_BIO_free(bio) }
+        guard i2d_OCSP_REQUEST_bio(bio, request) > 0 else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        // Copy from bio to byte array then convert to Data
+        var count = 0
+        var out: UnsafePointer<UInt8>?
+        guard CCryptoBoringSSL_BIO_mem_contents(bio, &out, &count) > 0 else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        let requestData = Data(UnsafeBufferPointer(start: out, count: count))
+
+        // X509_STORE is freed up after all OCSP_basic_verify calls finish.
+        // Don't do defer-free here because it would get freed up. `withExtendedLifetime(anchorCerts)` doesn't help.
```

Comment:
Not sure why this is. It happens for sk_x509 as well but we don't need it for `OCSP_basic_verify`. "free" is called on L389.

Reply:
It's not _entirely_ clear to me that this use of `X509_STORE` is thread-safe. Is it going to be better to construct one per invocation?

Reply:
Changed to create a new `X509_STORE` each time (https://github.com/apple/swift-package-manager/pull/3272/commits/3514d2c8226b768bf714eb8edd679b808adb5d62)